pkg/server/plugin/upstreamca/spire/spire_node.go

```diff
@@ -0,0 +1,241 @@
+
+	stream, err := nodeClient.FetchX509SVID(ctx)
+	if err != nil {
+		fmt.Println("error in fetch node stream [%s]")
+		return err
+	}
+	err = stream.Send(&nodeRequest)
+	if err != nil {
+		return err
+	}
+
+	stream.CloseSend()
+
+	var nodeResponse *node.FetchX509SVIDResponse
+	nodeResponse, err = stream.Recv()
+	if err != nil {
+		return err
+	}
+
+	m.intermediateCA, m.bundle, err = m.getCertFromResponse(nodeResponse)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *spirePlugin) submitCSRUpstreamCA (ctx context.Context, config *Configuration, csr []byte) error {
+	//	TBD :
+	//	get node connection
+	//	create csr
+	//	send  csr with CA option enabled
+	//	return the intermediate certificates (intermediate CA , intermediate key and trust bundle
+	fmt.Println("[plugin][upstreamca] get intermediate certificates using node api")
+	nodeClient, err := m.newNodeAPIClient(ctx, config)
+	if err != nil {
+		fmt.Println("error seen due to node client create error")
+		return err
+	}
+
+	csrs := [][]byte{}
+	csrs = append(csrs,csr)
+	nodeRequest := node.FetchX509SVIDRequest{Csrs:csrs}
+
+	stream, err := nodeClient.FetchX509SVID(ctx)
+	if err != nil {
+		fmt.Println("error in fetch node stream [%s]")
+		return err
+	}
+	err = stream.Send(&nodeRequest)
+	if err != nil {
+		return err
+	}
+
+	stream.CloseSend()
+
+	var nodeResponse *node.FetchX509SVIDResponse
+	nodeResponse, err = stream.Recv()
+	if err != nil {
+		return err
+	}
+
+	m.intermediateCA, m.bundle, err = m.getCertFromResponse(nodeResponse)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *spirePlugin) getCertFromResponse(response *node.FetchX509SVIDResponse) ([]*x509.Certificate, *bundleutil.Bundle, error) {
+	var err error
+	var svid []*x509.Certificate
+	var bundle *bundleutil.Bundle
+
+	if response.SvidUpdate == nil {
+		return nil, nil, errors.New("response missing svid update")
+	}
+	if len(response.SvidUpdate.Svids) < 1 {
+		return nil, nil, errors.New("no svid received")
+	}
+
+	svidMsg, ok := response.SvidUpdate.Svids[m.trustDomain.String()]
+	if !ok {
+		return nil, nil, fmt.Errorf("incorrect svid: %s", m.spiffeID)
+	}
+
+	svid, err = x509.ParseCertificates(svidMsg.CertChain)
+	if err != nil {
+		return nil, nil, fmt.Errorf("invalid svid: %v", err)
+	}
+
+	if response.SvidUpdate.Bundles == nil {
+		return nil, nil, errors.New("missing bundles")
+	}
+
+	fmt.Println("trust domain [%v]",m.trustDomain.String())
+	bundleProto := response.SvidUpdate.Bundles[m.trustDomain.String()]
+	if bundleProto == nil {
+		return nil, nil, errors.New("missing bundle")
+	}
+
+	bundle, err = bundleutil.BundleFromProto(bundleProto)
+	if err != nil {
+		return nil, nil, fmt.Errorf("invalid bundle: %v", err)
+	}
+
+	return svid, bundle, nil
+}
+
+func (m *spirePlugin) createCSRRequest(spiffeID string) ([]byte, error) {
+	key, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return nil, err
+	}
+
+	m.privateKey = key
```

Comment:
I'm not sure I see why this needs to be stored on the struct? The less we can store there the better I think

Reply:
Removed the redundant code .